vowpalwabbit/io_buf.h

```diff
@@ -10,54 +10,105 @@
 #include <vector>
 #include <memory>
 #include <cassert>
+#include <cstdlib>
 
 #include "v_array.h"
 #include "hash.h"
-#include "vw_exception.h"
 #include "io/io_adapter.h"
 
+#ifndef VW_NOEXCEPT
+#  include "vw_exception.h"
+#endif
+
 /* The i/o buffer can be conceptualized as an array below:
-**  _______________________________________________________________________________________
-** |__________|__________|__________|__________|__________|__________|__________|__________|   **
-** space.begin           space.head             space.end                       space.endarray **
+**  _________________________________________________________________
+** |__________|__________|__________|__________|__________|__________|
+** _buffer._begin        head       _buffer._end                     _buffer._end_array
 **
-** space.begin     = the beginning of the loaded values in the buffer
-** space.head      = the end of the last-read point in the buffer
-** space.end       = the end of the loaded values from file
-** space.endarray  = the end of the allocated space for the array
+** _buffer._begin     = the beginning of the loaded values in the buffer
+** head               = the end of the last-read point in the buffer
+** _buffer._end       = the end of the loaded values from file
+** _buffer._end_array = the end of the allocated _buffer for the array
 **
 ** The values are ordered so that:
-** space.begin <= space.head <= space.end <= space.endarray
+** _buffer._begin <= head <= _buffer._end <= _buffer._end_array
 **
-** Initially space.begin == space.head since no values have been read.
+** Initially _buffer._begin == head since no values have been read.
 **
-** The interval [space.head, space.end] may be shifted down to space.begin
+** The interval [head, _buffer._end] may be shifted down to _buffer._begin
 ** if the requested number of bytes to be read is larger than the interval size.
 ** This is done to avoid reallocating arrays as much as possible.
 */
 
 class io_buf
 {
+  struct internal_buffer
```

Comment:
do you think it is worth adding a comment here explaining that:

io_buf re-implements a grow-only variant of `v_array` in this `internal_buffer` since it needs to access the array's internals? Or maybe call `internal_buffer` `expand_only_v_array`?

Reply:
Added a comment explaining why this is here